nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+                })
+            }
+            #[inline]
+            fn from_pattern(_p: ()) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Complex type can't parse from ()"),
+                ))
+            }
+            #[inline]
+            fn conjugate(self) -> Result<Self, MMError> {
+                Ok(self.conj())
+            }
+            #[inline]
+            fn negative(self) -> Result<Self, MMError>
+            where
+                Self: Neg<Output = Self>,
+            {
+                Ok(-self)
+            }
+        }
+    };
+}
+/// Implement MMType for primitive unit types.
+macro_rules! mm_pattern_impl {
+    ($T:ty) => {
+        impl MMType for $T {
+            #[inline]
+            fn from_i128(_i: i128) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Pattern type can't parse from i128"),
+                ))
+            }
+            #[inline]
+            fn from_f64(_f: f64) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Pattern type can't parse from f64"),
+                ))
+            }
+            #[inline]
+            fn from_c64(_c: Complex<f64>) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Pattern type can't parse from Complex<f64>"),
+                ))
+            }
+            #[inline]
+            fn from_pattern(p: ()) -> Result<Self, MMError> {
+                Ok(p)
+            }
+
+            #[inline]
+            fn conjugate(self) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Pattern type has no conjugate"),
+                ))
+            }
+            #[inline]
+            fn negative(self) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Pattern type has no negative"),
+                ))
+            }
+        }
+    };
+}
+
+mm_int_impl!(i8);
+mm_int_impl!(i16);
+mm_int_impl!(i32);
+mm_int_impl!(i64);
+mm_int_impl!(i128);
+
+mm_float_impl!(f32);
+mm_float_impl!(f64);
+
+mm_complex_impl!(f32);
+mm_complex_impl!(f64);
+
+mm_pattern_impl!(());
+
+#[derive(Parser)]
+#[grammar = "io/matrix_market.pest"]
+struct MMParser;
+
+/// Parsing a pest structure to a Typecode of the matrix.
+fn parsing_header(inner: &mut Pairs<'_, Rule>, header_type: &mut Typecode) -> Result<(), MMError> {
+    // once the data can be parsed, all unwrap() in this function here will not panic.
+    header_type.sp = inner
```

Comment:
can you use full names for this, like `sparsity` instead of `sp`, and `data_type` instead of `dt` and so on?